### PR TITLE
Temporal simulator

### PR DESCRIPTION
**Pull request by Framba-Luca**
Addition of a TemporalSimulator

### CODE REVIEW

File: unified_planning/engines/temporal_simulator.py [hunk 1]
```diff
@@ -0,0 +1,941 @@
...
+        """IMPORTANT NOTE: This function modifies the data structures given as parameters."""
+        insert_interval(stn, last_event, event, left_bound=Fraction(0))
+        if event.kind == TemporalEventKind.START_ACTION:
+            committed_events: List[Event] = []
+            ce = event.committed_events
+            assert (
+                ce is not None
+            ), "Error, start Action event can't have None committed events"
+            for committed_event in ce:
+                if any(committed_event in el for el in running_events):
```
Comment (Framba-Luca) [inline unified_planning/engines/temporal_simulator.py:662]: Here I am thinking about a better way to find if an Event was already submitted in this "states chain".

The best idea I had so far is to hold a set in the state with all the events already applied. It might also be a `set[int]`, with only the events id's, if it helps.

File: unified_planning/engines/temporal_simulator.py [hunk 2]
```diff
@@ -0,0 +1,980 @@
...
+    if left_bound is not None:
+        stn.add(left_event, right_event, -left_bound)
+    if right_bound is not None:
+        stn.add(right_event, left_event, right_bound)
+
+
+def correct_order_events_to_apply(
+    events: Iterable[TemporalEvent], running_events: List[List[TemporalEvent]]
+) -> Iterator[Optional[TemporalEvent]]:
+    # TODO find a significative name
```
Comment (alvalentini) [inline unified_planning/engines/temporal_simulator.py:944]: Leftover

**Review by alvalentini — CHANGES_REQUESTED** [review]
@Framba-Luca as a general comment, the temporal simulator implementation is a bit hard to follow. Maybe a bit of refactoring (adding a few comments) could be useful.

File: unified_planning/engines/mixins/simulator.py [hunk 3]
```diff
@@ -235,9 +267,100 @@ def get_unsatisfied_goals(
...
+                            updated_values[fluent] = em.auto_promote(
+                                f_eval.constant_value() - v_eval.constant_value()
+                            )[0]
+                        else:
+                            raise NotImplementedError
+            if ev.simulated_effect is not None:
+                red_fluents.update(all_possible_assignments)
+                for f, v in zip(
+                    ev.simulated_effect.fluents,
+                    ev.simulated_effect.function(self._problem, state, {}),
```
[the remark this answers is not in the document]
Comment (Framba-Luca) [reply unified_planning/engines/mixins/simulator.py:359]: Because all the events are obtained from grounded actions.

[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/engines/mixins/simulator.py:359]: This is true for our implementations, but the `SimulatorMixin` should be more general. Right?

[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/engines/mixins/simulator.py:359]: After discussing with @mikand, we decided that extending the Event class with a reference to its Action and its actual parameters could be a good idea to solve this comment.

File: unified_planning/model/state.py [hunk 4]
```diff
@@ -36,7 +37,14 @@ class COWState(ROState):
...
     def make_child(
-        self, updated_values: Dict["up.model.FNode", "up.model.FNode"]
+        self,
+        updated_values: Dict["up.model.FNode", "up.model.FNode"],
+        running_events: Optional[
+            List[List["up.engines.mixins.simulator.Event"]]
+        ] = None,
+        stn: Optional["up.model.delta_stn.DeltaSimpleTemporalNetwork"] = None,
+        durative_conditions: Optional[List["up.model.fnode.FNode"]] = None,
+        events: Optional[Set["up.engines.mixins.simulator.Event"]] = None,
```
Comment (alvalentini) [inline unified_planning/model/state.py:47]: This new parameters are not documented and in general I don't think we want to introduce temporal state in this way.

Probably we want a `ROTemporalState`, a `COWTemporalState` and a `UPCOWTemporalState`.
